src/cast_core/src/ExternalScriptRunner.cpp

```diff
@@ -0,0 +1,138 @@
+    std::string script;
+
+    std::string mongoServerURI;
+
+    // Record data retruned by the script
+    metrics::Operation operation;
+
+    // Record data of the script operation itself
+    metrics::Operation scriptOperation;
+
+    PhaseConfig(PhaseContext& phaseContext, ActorId id)
+        : script{phaseContext["Script"].to<std::string>()},
+          mongoServerURI{phaseContext["MongoServerURI"].maybe<std::string>().value_or("--nodb")},
+          operation{phaseContext.operation("DefaultMetricsName", id)},
+          scriptOperation{phaseContext.namedOperation("ExternalScript", id)} {}
+};
+
+/**
+ * @brief Execute external script
+ * 
+ * @param cmd 
+ * @return std::string 
+ */
+std::string ExternalScriptRunner::exec(const char* cmd) {
+    std::array<char, 128> buffer;
+    std::string result;
+    std::unique_ptr<FILE, decltype(&pclose)> pipe(popen(cmd, "r"), pclose);
+    if (!pipe) {
+        throw std::runtime_error("popen() failed!");
+    }
+    while (fgets(buffer.data(), buffer.size(), pipe.get()) != nullptr) {
+        result += buffer.data();
+    }
+    return result;
+}
+
+void ExternalScriptRunner::run() {
+    // This section gets executed before the phases. Setup is executed here, so maybe
+    // we could add some sanity checks before actually running the shell.
+
+    for (auto&& config : _loop) {
+
+        std::string programCommand = "";
+        
+        if (_command == "mongosh") {
+            programCommand = "mongosh " + config->mongoServerURI + " --quiet --eval \"" + config->script + "\"";
+        }
+        else if (_command == "sh") {
+            programCommand = "sh -c \"" + config->script + "\"";
+        }
+        else {
+            throw std::runtime_error("Script type " + _command + " is not supported.");
+        }
+        
+        BOOST_LOG_TRIVIAL(info) << "ExternalScriptRunner running "
+                                << programCommand;
+
+        for (const auto&& _ : config) {
+
+            // Execute the script and read result from stdout
+            auto ctx = config->scriptOperation.start();
+            const char* programCmdPtr = &*programCommand.begin();
+            std::string result = exec(programCmdPtr);
+            ctx.success();
+
+            try {
+
+                // If the result from stdout can be parsed as integer, report the operation metrics
+                // Otherwise, ignore the result
+                int num = std::stoi(result);
+                config->operation.report(metrics::clock::now(),
+                                              std::chrono::milliseconds{num}); 
+            }
+            catch(std::exception& err) {
+                BOOST_LOG_TRIVIAL(info) << err.what();
+            }
+        }
+    }
+}
+
+ExternalScriptRunner::ExternalScriptRunner(genny::ActorContext& context)
+    // These are the attributes for the actor.
+    : Actor{context},
+      _command{std::move(context.get("Command").maybe<std::string>().value_or(""))},
```

Comment:
This command might make more sense in the `PhaseConfig` so you can access `config->command` (`config` is your `PhaseConfig`). See also other comment about constructing the command in there as well.